Issue 329 - Adding Koa2 support

jpravetz opened the pull request:
All existing unit tests running EXCEPT for access unit tests. I haven't started on any koa/* tests. Input and edits most welcome. I still haven't completely got my head around ERMObject, thinking it introduces complexity and is perhaps overkill compared to a plain object. But if it works, okay by me.

jpravetz committed:
Adding koa2 support.

jpravetz committed:
Added switch to specify which middleware to load

jpravetz committed:
Added universal context object

jpravetz committed:
Refactored context with inheritance,

jpravetz committed:
Fix lint issues. Fix some issues in Context.js. Move getPostMiddlewareForMethods to shared.js. Start unit testing (test-unit yields 105 passing, 4 failing)

jpravetz committed:
test-unit working (that’s not all unit tests)

jpravetz committed:
Fix typo preventing test-ci from getting far. Now it gets far, but has errors still.

jpravetz committed:
Get express.createDocuments routes unit tests working

jpravetz committed:
Tighten up some code bloat

jpravetz committed:
Fixed express read, update, delete unit tests

b-gran commented:
That's a reasonable point about ERMOperation. There should probably be documentation in the code base explaining the rationale:

Before ERMOperation, every user-supplied middleware had full access to ERM's private state. Every time a middleware hook was called was an opportunity for ERM's private state to be completely trashed, and writing safe middleware meant checking for type/existent of state properties *every* time you wanted to access them. 

ERMOperation provides runtime-guarantees about 
1. The type of state properties
2. The existence of state properties 
3. That state relied upon by prior middleware is unmodified 

This is even more important for Koa middleware since middleware at the top of the stack can pass control to other middleware and then resume operation. 

It also acts as an adapter between request-coupled legacy middleware and request-independent middleware. Maybe it makes to split the immutability code and the adapter into separate files.

jpravetz commented:
I'm wondering if this really matters, given that's pretty much the norm in javascript, which is part of what makes it easy. Aren't these issues true also of the ctx, req, res objects? Also, it would be helpful if you built ImmutableObject into a non minified version for nodejs.  Anyway, up to you, as I feel unqualified to provide deeper insight. I like what you've done with the repo.

jpravetz committed:
All unit tests are working. Koa unit tests are not yet written. Would appreciate review of properties that have been placed in req.erm and req._erm (express).

jpravetz committed:
Starting on koa unit tests. Should be able to reuse integration tests.

jpravetz committed:
Moved koa options to options.koa.router and options.koa.compose. This way non-koa apps do not need to require these modules

jpravetz committed:
Working on initialization for koa unit tests

jpravetz committed:
Modify again how koa app is initialized for testing

jpravetz commented:
@b-gran I'm having trouble figuring out where the request body gets transferred into state.body. The createObject middleware (as an example) uses state.body rather than req.body, so it must be being set somewhere?

b-gran commented:
It gets added in the (somewhat badly named) [filterRequestBody](https://github.com/b-gran/express-restify-mongoose/blob/issue-305/src/api/filterRequestBody.js#L25) transformation.

jpravetz commented:
Oh yes, of course. I'd seen that before. Thanks.

jpravetz committed:
Got some unit tests working for koa. Fixed koa outputFn so it is called, but it must now always return a promise. Added debug package. Fixed bug in Transformation for koa. Fixed outputFn incorrect return of next(). Only call postMiddleware now if set.

jpravetz committed:
Fixing unit tests and bugs that they are uncovering. express and restify unit tests still running. Down to 83 koa failures.

jpravetz committed:
Introduced koa-qs query string parser, needed for nested query parameters. Fixed koa access middleware to handle (ctx,done) functions as is being used by unit tests.
Still working on koa routing issues that are affecting some unit tests.

jpravetz committed:
Added debug statements to express. Made more koa functions non-anonymous for easier middleware chain debugging.

jpravetz committed:
Fixed koa access middleware and another 40 unit tests. 42 koa unit tests to still fix.

jpravetz committed:
Removed compose from options.postProcess. Got middleware unit tests to work with express and koa (required koa/express switches throughout middleware tests). Started adding koa-specific unit tests, and currently working on access unit test.

jpravetz committed:
Got koa unit tests for ensureContent, onError, outputFn to work

jpravetz committed:
Got remaining koa-unit unit tests working and fixed regression issues with express middleware unit tests

jpravetz committed:
Fix lint issues, verify that all unit tests run that were not already broken because of dtrace dependencies.

jpravetz committed:
Now that tests are working, I’m having a closer look at koa routes. I don’t believe I’ve got things set up entirely correctly. Will need some more work. Also refactored the code to output results and handle errors, and updated documentation.

jpravetz committed:
Switched from koa-router to koa-better-router for unit tests. koa-router seemed to be creating incorrect middleware chains, building chains from all matching route paths.
Added initRoute to middleware chain for all get/put/post/del methods rather than using app.use. This was to prevent the same middleware from being added more than once by more than one call to restify function, as was being done in integration tests (eg. create). This has fixed up middleware chains for koa.

jpravetz committed:
Fixed unit test failure when postCreate throws. This will result in 400, but erm.statusCode will have already been set to 201, which I think is fine since it’s not returned.

jpravetz committed:
Finish integration tests for koa using both koa-router and koa-better-router.
Modify integration tests so there is only one koa and router context per describe function.
Set koa-router to point to patched version of koa-router.
